### PR TITLE
fix message links with unbox query

### PR DESCRIPTION
**Pull request by ahdinosaur**
the existing test case didn't include an edge case:

if the unbox key includes a '+' character, when parsed as a query string the '+' becomes a ' ' character, which is no longer a correct unbox key.

this changes the test to match the real-world test case from https://github.com/ssbc/secure-scuttlebutt/pull/220 and fixes the behavior with a hack.

### CODE REVIEW

**Comment by christianbundy** [comment]
Nice catch! Can a query string have multiple `+` characters? If so, this may only be replacing the first search result.

**Comment by ahdinosaur** [comment]
> Can a query string have multiple + characters?

@christianbundy yes, a '+' character has no special significance in the base64-encoded unbox key, but it does have semantic meaning in a query string (it means the ' ' character). so yes we want to replace all occurrences, i didn't realize `string.replace` didn't do that.

**Comment by ahdinosaur** [comment]
this is also a total hack, but we deserve the hack from using url unsafe characters and pretending it's a url. i'm really keen in any future protocol changes, if we expect to use something as a human-readable identifier, it should be encoded in [base58](https://en.wikipedia.org/wiki/Base58), so we avoid using non-alphanumeric characters (like '+' and '/' which we use now, but also '~' and '.' which are technically url safe but why take that risk).

**Comment by christianbundy** [comment]
Yeah, I think [`String.replace()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace) can be a bit of a footgun for that reason. I'll make an inline comment, not sure why I didn't start the conversation there in the first place.

I'm *super* excited at the idea of using real URLs for ssb:// content, there are lots of great solutions.

**Comment by dominictarr** [comment]
if you use a regular expression with `g` mode enabled, it will replace every occurance. as I did here: https://github.com/ssbc/secure-scuttlebutt/blob/b758c09655ae71d2e1bbc295842694425134ff41/index.js#L84

File: index.js [hunk 1]
```diff
@@ -171,7 +171,10 @@ exports.parseLink = function parseBlob (ref) {
   var match = parseLinkRegex.exec(ref)
   if (match && match[1]) {
     if (match[3]) {
-      return {link: match[1], query: Querystring.parse(match[4])}
+      var query = Querystring.parse(match[4])
+      // unbox keys have a '+' in them that is parsed into a ' ', this changes it back
+      if (query.unbox) query.unbox = query.unbox.replace(' ', '+')
```
Comment (christianbundy) [inline index.js]: For multiple replacements I think we'd need something like:

```js
if (query.unbox) query.unbox = query.unbox.replace(/ /g, '+')
```

There's another funky edge-case here where `unbox` could be a truthy non-string (like `42`), and `(42).replace()` throws a nasty error. Maybe something like this?

```js
if (isString(query.unbox)) query.unbox = query.unbox.replace(/ /g, '+')
```

Could be overkill.

Comment (dominictarr) [reply index.js]: if unbox isn't a base64 for a 32 byte key, it's not a valid key, so it should be ignored in that case.

**Comment by mmckegg** [comment]
Okay, I'm merging this!

**Comment by mmckegg** [comment]
📦  ssb-ref@2.13.8
